lms/djangoapps/course_api/api.py

```diff
@@ -0,0 +1,55 @@
+"""
+Course API
+"""
+
+from django.contrib.auth.models import User
+from rest_framework.exceptions import PermissionDenied
+
+from lms.djangoapps.courseware.courses import get_courses, get_course_with_access
+
+from .permissions import can_view_courses_for_username
+from .serializers import CourseSerializer
+
+
+def get_effective_user(requesting_user, target_username):
+    """
+    Get the user we want to view information on behalf of.
+    """
+    if target_username == requesting_user.username or None:
+        return requesting_user
+    elif can_view_courses_for_username(requesting_user, target_username):
+        return User.objects.get(username=target_username)
+    else:
+        raise PermissionDenied()
+
+
+def course_detail(requesting_user, username, course_key, request):
```

Comment:
Because I want the API's interface to explain what the API does as much as possible.  The contract of the API is "for the given user, asking on behalf of the target_username, give me information about the course identified by course_key.  The `request` argument is there because we need it to get the course description, but that's due to external constraints.  The primary interface is requesting_user, username, course_key.  I think taking the user off the request object sacrifices too much clarity for a small amount of simplicity.  

If we wanted to pull the `requesting_user` from the `request` object, we could also pull `target_username` and the `course_key` off the request object.  That information is all available, but `course_detail(request)` is a terrible api signature, that doesn't tell the user what information the API needs.  

Later, when we extract out the user-masquerading code from the API, the signature will just be `course_detail(user, course_key, request)`, and if we put the course description on the CourseOverview, it will just be `course_detail(user, course_key)`.  But for any version of it, we're still telling the API "who" and "what."